phy: efinix: titanium: add lvds 1000base-x

maass-hamburg opened the pull request:
uses 1 lvds tx and 4 lvds rx of the titanium (not using the Titanium transceivers)

needs https://github.com/enjoy-digital/litex/pull/2192


for testing: https://github.com/litex-hub/litex-boards/pull/655

m-byte commented on liteeth/phy/titanium_1000basex.py:
```diff
@@ -0,0 +1,431 @@
+class EfinixSerdesBuffer(LiteXModule):
+    def __init__(self, data_in, data_in_len, data_out, data_out_valid, allign):
+
+        self.alligner = alligner = EfinixAlligner(allign)
```
```suggestion
        self.aligner = aligner = EfinixAligner(align)
```

maass-hamburg committed:
phy: pcs: send config to sgmii phy according to standard

according to the sgmii standard, when the tx_config_reg is
send from the mac to the phy, only bit 0 and 14 should be
set to 1, the other bits should be set to 0, as they are
reserved.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
phy: pcs: add options for CSRs

This patch adds options for the PCS CSRs to the phy driver.
With this a phy driver can be implemented that can detect
the status of the link and react to it.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
phy: efinix titanium lvds 1000basex: add option for rx termination

Be able to select between internal and external termination for the
RX side of the 1000Base-X PHY. Using the internal ones is not quite right by spec,
because we are using 4 100 Ohm resistors in parallel restulting in 25 Ohm, instead of just one 100 Ohm resistor. Just activating one internal termination resistor lead to a timing problem.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
phy: efinix titanium lvds 1000basex: activate CSRs from PCS

activate CSRs from PCS

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
phy: pcs: improve and prepare it

improve the 1000basex pcs and prepare it
by adding a input valid signal

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
phy: add efinix titanium 1000base-x lvds

add a phy for the efinix titanium, that just uses lvds
ports as io.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
tests: phy: add test/sim for efinix 1000basex lvds phy

add test/sim for efinix 1000basex lvds phy

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
phy: efinix titanium lvds 1000basex: decrease buffer size

decrease buffer size.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
phy: efinix titanium lvds 1000basex: add option for i2c

add option for i2c

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
phy: efinix titanium lvds 1000basex: add restart

add restart

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
phy: pcs: add delay for interrupt

make sure that the iface is really up stabiliy,
before raising a interrupt.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

enjoy-digital reviewed (approved):
Thanks @maass-hamburg, I'm going to merge and will push some changes on top of it.

enjoy-digital commented:
This is now merged and reworked a bit with the following commits:
- https://github.com/enjoy-digital/liteeth/commit/209bbb2e9473df2c2f2143baf6a5896f30433747
- https://github.com/enjoy-digital/liteeth/commit/5667738884cee8a888125e37217f480ca3c2efea
- https://github.com/enjoy-digital/liteeth/commit/74c6526d80ea4beeec108a809fdd2330082161e0
- https://github.com/enjoy-digital/liteeth/commit/8aaa6cb7fa4fa1a3951e6f44f4294f16a633a203

These commits switch the code to the Synchronous 8B10B Decoder and simplify/cleanup the code a bit.

maass-hamburg commented:
> This is now merged and reworked a bit with the following commits:
> - https://github.com/enjoy-digital/liteeth/commit/209bbb2e9473df2c2f2143baf6a5896f30433747
> - https://github.com/enjoy-digital/liteeth/commit/5667738884cee8a888125e37217f480ca3c2efea
> - https://github.com/enjoy-digital/liteeth/commit/74c6526d80ea4beeec108a809fdd2330082161e0
> - https://github.com/enjoy-digital/liteeth/commit/8aaa6cb7fa4fa1a3951e6f44f4294f16a633a203
> 
> These commits switch the code to the Synchronous 8B10B Decoder and simplify/cleanup the code a bit.

Why was 74c6526d80ea4beeec108a809fdd2330082161e0 done? It is needed by this PR. At least for this implementation it is 100% necessary, but it depends on the used hardware.

enjoy-digital commented:
I would like to avoid code becoming too complex, the PCS code is expected to run at RX / TX rate and haven't found in the code where input_valid was used. If used by this implementation, can you explain the purpose and provide more info?

maass-hamburg commented:
> I would like to avoid code becoming too complex, the PCS code is expected to run at RX / TX rate and haven't found in the code where input_valid was used. If used by this implementation, can you explain the purpose and provide more info?

This uses it. Here we can't exactly recover the clock from the signal (clock recovery), like the other can do with their hard ip. So here we do it in logic. Normally we would get 10 bits per cycle, but to compensate for the frequency offset between the fpga and the endpoint, we can also get just 9 or 11 points per cycle. We have a buffer in between to make sure to always return 10 points or none. While none the signal will not be valid. I used the DecoderComb, so I don't had to delay that valid signal too.

enjoy-digital commented:
OK thanks, I'll do another review, but we should try to be as close as possible to other implementation, so if required, simplify input_valid and prefer delaying this signal instead of having a specific implementation of the Decoder.

maass-hamburg commented:
Also 5667738884cee8a888125e37217f480ca3c2efea should not have happened. The logic around it is time critical and we expected the result in the same cycle and not one later. This is from the logic, that detects an I2, which means we could remove it and by doing that empty our buffer by 20 points, if possible.

maass-hamburg commented:
> OK thanks, I'll do another review, but we should try to be as close as possible to other implementation, so if required, simplify input_valid and prefer delaying this signal instead of having a specific implementation of the Decoder.

Then we just have to delay the input valid by one cycle before it's used by the fsm

enjoy-digital commented:
OK thanks for the clarification. The original code wouldn't have been easy to maintain, so I'm trying to find a compromise.  We'll converge to a solution that will satisfy each other and just need to discuss things calmly and progress on each points.

maass-hamburg commented:
> > I would like to avoid code becoming too complex, the PCS code is expected to run at RX / TX rate and haven't found in the code where input_valid was used. If used by this implementation, can you explain the purpose and provide more info?
> 
> This uses it. Here we can't exactly recover the clock from the signal (clock recovery), like the other can do with their hard ip. So here we do it in logic. Normally we would get 10 bits per cycle, but to compensate for the frequency offset between the fpga and the endpoint, we can also get just 9 or 11 points per cycle. We have a buffer in between to make sure to always return 10 points or none. While none the signal will not be valid. I used the DecoderComb, so I don't had to delay that valid signal too.

This is still not fully fixed, 35cde07d871b6a9afc5e7b57682bd34b974c5daa is not enough, we still need the data_valid it in the fsm and the other places where it got removed.

enjoy-digital commented:
@maass-hamburg: Do you see a regression with this on hardware?

maass-hamburg commented:
> @maass-hamburg: Do you see a regression with this on hardware?

No, because it is not easy to test and I know it's still wrong.

With this we will duplicate the former symbol, if ce is not true, that will lead to a malformed packed, because one byte in in it was duplicated. Instead we should not be allowed to work with the duplicated symbol/byte.

maass-hamburg commented:
I would like to propose https://github.com/enjoy-digital/liteeth/pull/190

enjoy-digital commented:
Thanks @maass-hamburg, can you have a look at https://github.com/enjoy-digital/liteeth/commit/3b6002236864022b778dca3735a5272fe623748e? It should be similar functionally  to your PR (just a cycle of latency of difference, but should not impact things) and is simpler.

enjoy-digital commented:
@maass-hamburg: Does it look goog to you? If so, we could probably close https://github.com/enjoy-digital/liteeth/pull/190 and https://github.com/enjoy-digital/litex/pull/2246.

maass-hamburg commented:
@enjoy-digital seems working

enjoy-digital commented:
Thanks @maass-hamburg for the feedback.
